Hawkbit add custom device attributes

burumaj opened the pull request:
Adds method for setting custom device attributes that are provided
to the hawkBit server. These attributes are set by conditionally
including a header that can provide an implementation of the required
function, config struct and json_description.

burumaj commented:
This MR addresses  #37640

ycsin commented on include/mgmt/hawkbit/config_data.h:
```diff
@@ -0,0 +1,19 @@
+ */
+#ifndef ZEPHYR_INCLUDE_MGMT_HAWKBIT_CONFIG_DATA_H_
+#define ZEPHYR_INCLUDE_MGMT_HAWKBIT_CONFIG_DATA_H_
+#include <autoconf.h>
```
I think `autoconf.h` is automatically included?

burumaj replied:
Should be addressed now

ycsin commented on include/mgmt/hawkbit/config_data.h:
```diff
@@ -0,0 +1,19 @@
+ * should be set on.
+ * @return 0 on succes, otherwise negative error code
+ */
+int hawkbit_get_config_data(struct hawkbit_cfg_data *data);
```
this function should be in the existing `hawkbit.h` public header, or the public header should be moved into the 
new `hawkbit` folder

Navin-Sankar replied:
as @ycsin said `hawkbit_get_config_data` function should be placed in the existing public header `hawkbit.h`

burumaj replied:
My reasoning for putting it in a seperate header was this being a function that is internal to hawkBit. With my idea being that `hawkbit.h` would contain the public hawkbit API, and headers located in `subsys/mgmt/hawkbit/` would include optionally implemented functions, like in this case `hawkbit_get_config_data`, that are internally in hawkbit.

Apart from the above mentioned consideration there is also a technical one. `hawkbit_get_config_data` needs the definition of `struct hawkbit_config_data` in scope. This struct can defined in two ways depending on `CONFIG_HAWKBIT_CUSTOM_CONFIG_DATA`:

1. The default implementation is used. In this case (`CONFIG_HAWKBIT_CUSTOM_CONFIG_DATA=n`) `struct hawkbit_config_data` is defined in `zephyr/subsys/hawkbit/mgmt/hawkbit_config_data.h`. Only after this struct is defined, is the definition of `hawkbit_get_config_data` included because it depends on this struct having a definition, otherwise you get a compiler error.
2. The case where `CONFIG_HAWKBIT_CUSTOM_CONFIG_DATA=y`. Now the definition of `struct hawkbit_config_data` is in `include/mgmt/hawkbit/hawkbit_custom_data.h`, again this is now included `zephyr/subsys/hawkbit/mgmt/hawkbit_config_data.h`, before it includes the definition of `hawkbit_get_config_data`. Doing it this way prevents having to include `include/mgmt/hawkbit/hawkbit_custom_data.h` halfway through hawkbit.c.

This is the way I see it at least. Feel free to correct my way of thinking if I've got it wrong

ycsin replied:
Sounds a bit complicated, my limited and simple understanding is that headers are considered public when they are in the `include` folder and are supposed to be included by user application to interact with the system applications. Real private headers should be within the same folder of `hawkbit.c` instead.

ycsin replied:
would love to have your comments @gmarull

ycsin commented on subsys/mgmt/hawkbit/hawkbit_config_data.c:
```diff
@@ -0,0 +1,35 @@
+	return -ENOSYS;
+}
+#else
+
```
remove unnecessary empty line

burumaj replied:
Should be addressed now

ycsin commented on subsys/mgmt/hawkbit/hawkbit_config_data.c:
```diff
@@ -0,0 +1,35 @@
+#include <errno.h>
+#include "hawkbit_config_data.h"
+
+
```
remove unnecessary empty line

burumaj replied:
Should be addressed now

ycsin commented on subsys/mgmt/hawkbit/hawkbit_config_data.h:
```diff
@@ -0,0 +1,31 @@
+	const char *hwRevision;
+};
+
+static const struct json_obj_descr json_cfg_data_descr[] = {
```
I think variable should not be in header

ycsin commented on subsys/mgmt/hawkbit/hawkbit_config_data.h:
```diff
@@ -0,0 +1,31 @@
+#define __HAWKBIT_CONFIG_DATA_H__
+
+#include <data/json.h>
+#include <autoconf.h>
```
I think `autoconf.h` should not be included

burumaj replied:
Should be addressed now

burumaj committed:
mgmt: hawkbit: add setting of custom device attributes

Adds method for setting custom device attributes that are provided
to the hawkBit server. These attributes are set by conditionally
including a header that can provide an implementation of the required
function, config struct and json_description.

Signed-off-by: Joep Buruma <burumaj50@gmail.com>

burumaj committed:
samples/subsys/mgmt/hawkbit: add custom attributes to sample

Include custom device attributes in sample. And add a test to
twister to ensure that this new configuration builds.

Signed-off-by: Joep Buruma <burumaj50@gmail.com>

ycsin commented on subsys/mgmt/hawkbit/hawkbit_priv.h:
```diff
@@ -14,6 +14,7 @@
 #define __HAWKBIT_PRIV_H__
 
 #include <data/json.h>
+#include "hawkbit_config_data.h"
```
I'm not sure if a system application is allowed to include a header from the user application

I'm not sure who to ask though, @gmarull you have any idea about this?

ycsin commented on subsys/mgmt/hawkbit/hawkbit_config_data.c:
```diff
@@ -0,0 +1,33 @@
+#endif
+
+#if defined(CONFIG_HAWKBIT_CUSTOM_CONFIG_DATA)
+int __attribute__((__weak__)) hawkbit_get_config_data(struct hawkbit_cfg_data *data)
```
should be just `__weak`

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -1326,6 +1333,7 @@ enum hawkbit_response hawkbit_probe(void)
 free_md:
 	mbedtls_md_free(&hb_context.dl.hash_ctx);
 
+
```
unnecessary empty line

burumaj commented:
I don't have much time to work on this at the moment. I expect I will be able to get to this is next couple weeks.
